### DISCUSSION TITLE
Help with Error Messages Using Attach/Assign Actions

### DISCUSSION THREAD

**Discussion by willardinteractive** [filamentphp/filament#6137] [Help] [answered]
I am struggling with using the Attach and Assignment actions on Relation Managers. I'm sure if I am trying to use this outside of the intent so I'm hoping someone has some insight here.

I have two models I am working with Locations and Companies. Each Company can have many locations.

My Company model file has this relationship defined:

```
public function locations(): HasMany
{
     return $this->hasMany(Location::class);
}
```

My Location model file has this relationship defined:
```
public function company(): BelongsTo
{
     return $this->belongsTo(Company::class);
}
```

On my add/edit screen for managing Locations I have a searchable select field to tie the location to a specific company. This works fine.

```
Select::make('company_id')
   ->relationship('company', 'name')
   ->searchable()
   ->required(),
```

For the CompanyResource I created a relation manager for showing locations.

When I visit the edit screen for a Company entry I see the correct Locations listed in the Relation Manager. This seems to be working correctly as well.

![image](https://user-images.githubusercontent.com/826203/230435294-57e61f1c-68aa-4d81-9c5c-440f85a37501.png)

I'd like to be able to (I'm not sure which is the correct one here) 'attach' or 'assign' existing Locations through this Relation Manager. I'm not if that's possible or not but looking at the documentation I'm guessing that the AssociateAction is what I want since this is on the CompanyResource, which hasMany LocationResources. Is that correct?

My CompanyResource/RelationManagers/LocationsRelationManager.php has this action:

```
->headerActions([
        Tables\Actions\CreateAction::make(),
        Tables\Actions\AssociateAction::make(),
    ])
```

As you can see in the screenshot above, I do get the associate action button. When I click it I get the modal.

![image](https://user-images.githubusercontent.com/826203/230435598-46cc32eb-d171-4198-9180-f6d28f97974a.png)

After I begin typing to search for a Location to associate I get the following error:

`Call to undefined method App\Models\Location::companies()`

This seems backwards. The Location model only has a company() method.

I thought that I might be misunderstanding when Attach and Associate are applicable so I tried to use AttachAction as well. This returns the following error when I begin to search for a Location in the modal.

`Call to undefined method Illuminate\Database\Eloquent\Relations\HasMany::getRelatedKeyName()`

That error seems more like Attach is the appropriate action I need to use but I am not clear on why I am getting this error.

Thanks in advance. Happy to provide more information if I am missing something important. I looked at other topics here but didn't see anything related to this specifically.

**Comment by danharrin** [comment, marked as answer]
Associate is correct here, as it's a one-to-many relationship. Try setting the `$inverseRelationship = 'company'` on the relation manager class

**Reply by willardinteractive** [reply]
Thank you so much! That worked! I added the following and it worked great!

`protected static ?string $inverseRelationship = 'company';`

**Reply by jdion84** [reply]
thank you this worked perfectly as i named my relation in my model differently than the standard way.

**Comment by classicalguss** [comment]
I got the same issue here. The solution in the answer worked. But this issue stayed with me for more than 2 hours, and yet it's a very common use-case to have HasMany, BelongsTo relation. I wonder if Associate is not meant to be used for this kind of relation? It seems to me that Filament should easily find the relation name. I am tempted to work on a PR, but I'm worried that I'm missing something here.
